### DISCUSSION TITLE
Help with Contributing on Windows

### DISCUSSION THREAD

**Discussion by GoGurtPrime** [timostamm/protobuf-ts#476] [Q&A] [answered]
I've contributed to the project before, and each time I have difficulty with building the project. According to the contributing markdown file `CONTRIBUTING.md`, I should be able to install the necessary dependencies and run `make`. 

The first issue I've run into is the non-Windows friendly syntax used for the commands found in the `Makefile`, which I've temporarily changed to `npm exec lerna ...` to allow the Lerna commands to work. So far so good, this seems to clear up the Windows specific build issues.

My main issue I'm running into is when the following command in the Makefile is executed:
`npm exec lerna bootstrap`

Here is the output of that command:

```
lerna notice cli v3.22.1
lerna info Bootstrapping 22 packages
lerna info Installing external dependencies
lerna ERR! npm install exited 1 in '@protobuf-ts/example-angular-app'
lerna ERR! npm install stderr:
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR!
npm ERR! While resolving: @protobuf-ts/example-angular-app@2.8.2
npm ERR! Found: @angular/compiler@10.0.14
npm ERR! node_modules/@angular/compiler
npm ERR!   @angular/compiler@"~10.0.6" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer @angular/compiler@"10.2.5" from @angular/compiler-cli@10.2.5
npm ERR! node_modules/@angular/compiler-cli
npm ERR!   dev @angular/compiler-cli@"^10.0.6" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
npm ERR! See C:\Users\JohnSmith\AppData\Local\npm-cache\eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\JohnSmith\AppData\Local\npm-cache\_logs\2023-01-26T18_02_20_380Z-debug-0.log

lerna ERR! npm install exited 1 in '@protobuf-ts/example-angular-app'
lerna WARN complete Waiting for 20 child processes to exit. CTRL-C to exit immediately.
make: *** [Makefile:13: lerna-bootstrap] Error 1
```

I'm not sure if this is caused by changing over the syntax to use `npm exec`, but I've also tried `npx lerna bootstrap` with no luck. I'm currently trying to upgrade the `@protobuf-ts/runtime-angular` to be compatible with Angular 15 so I can make a pull request.

-------------------------------------------------------------------------------------------------------------------------------------

**UPDATE:**

I've gone ahead and used Windows Subsystem for Linux to build with Ubuntu. Installation of packages worked this time, I believe using `npx` or `npm exec` may break options. I wanted to run the file exactly as it was in the repository.

I'm now past dependency installation, and I'm experiencing an issue in which the type `TextEncoder` does not exist. I'm really confused about the build process here. I'd like to familiarize myself to better contribute to this project, so any help here is greatly appreciated.

To clarify, this is occurring in `@protobuf-ts/runtime`, in which I receive the output `@protobuf-ts/runtime: Executed 623 of 623 specs (103 FAILED) in 0.409 sec.`

**Comment by timostamm** [comment, marked as answer]
Hey Colin, I don't think it's feasible to build this project on Windows without the WSL. Maybe the contributing.md should be clear about that, but the entire project setup could use an overhaul :) 

One obstacle you will probably hit is the conformance test runner. It is not available as a pre-built binary, so it needs to be [build from source](https://github.com/timostamm/protobuf-ts/blob/master/packages/test-generated/Makefile#L26). This requires `bazel` (I have since learned that it is best practice to use `bazelisk` to manage the bazel installation) and probably some standard build tools. I hope that a separate open source repository can take on the responsibility of creating a build of the conformance test runner and make it available as a downloaded release for several platforms.

If you look at .circleci/config.yml, you see that it uses a pre-built binary for linux x86 in CI to speed things up, so you might be able to use just that.

I'd love to use turborepo here instead of lerna in the long run.

**Reply by GoGurtPrime** [reply]
Tim, thanks for the helpful notes! I remember you mentioned difficulty building on Windows but couldn't find where you mentioned it in our old discussions. I'll take a look at the CI pipeline to familiarize myself with what's going on and try to use the pre-built binary. I appreciate the help, I have worked with many pipelines but never with tools like Lerna or Bazel.
